src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+        self._check_solver(solver_name)
+        self._solver_name = solver_name
+
+    @property
+    def solver_kwargs(self):
+        return self._solver_kwargs
+
+    @solver_kwargs.setter
+    def solver_kwargs(self, solver_kwargs: dict):
+        self._check_solver_kwargs(self.solver_name, solver_kwargs)
+        return self._solver_kwargs
+
+    def _check_solver(self, solver_name: str):
+        """
+        Ensure the provided solver name is allowed.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver to be checked.
+
+        Raises
+        ------
+        ValueError
+            If the provided solver name is not in the list of allowed optimizers.
+        """
+        if solver_name not in self.allowed_optimizers:
+            raise ValueError(
+                f"Solver `{solver_name}` not allowed for "
+                f"{self.__class__} regularization. "
+                f"Allowed solvers are {self.allowed_optimizers}."
+            )
+
+    @staticmethod
+    def _check_solver_kwargs(solver_name, solver_kwargs):
+        """
+        Check if provided solver keyword arguments are valid.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver.
+        solver_kwargs :
+            Additional keyword arguments for the solver.
+
+        Raises
+        ------
+        NameError
+            If any of the solver keyword arguments are not valid.
+        """
+        solver_args = inspect.getfullargspec(getattr(jaxopt, solver_name)).args
+        undefined_kwargs = set(solver_kwargs.keys()).difference(solver_args)
+        if undefined_kwargs:
+            raise NameError(
+                f"kwargs {undefined_kwargs} in solver_kwargs not a kwarg for jaxopt.{solver_name}!"
+            )
+
+    @staticmethod
+    def _check_is_callable_from_jax(func: Callable):
+        """
+        Check if the provided function is callable and from the jax namespace.
+
+        Ensures that the given function is not only callable, but also belongs to
+        the `jax` namespace, ensuring compatibility and safety when using jax-based
+        operations.
+
+        Parameters
+        ----------
+        func :
+            The function to check.
+
+        Raises
+        ------
+        TypeError
+            If the provided function is not callable.
+        ValueError
+            If the function does not belong to the `jax` or `neurostatslib.glm` namespaces.
+        """
+        if not callable(func):
+            raise TypeError("The loss function must a Callable!")
+
+        if (not hasattr(func, "__module__")) or (
+            not (
+                func.__module__.startswith("jax")
+                or func.__module__.startswith("neurostatslib.glm")
+            )
+        ):
+            raise ValueError(
+                f"The function {func.__name__} is not from the jax namespace. "
+                "Only functions from the jax namespace are allowed."
```

Comment:
doesn't the `hasattr(func, "__module__")` mean that things outside jax namespace are allowed?

Reply:
I removed all of that but to clarify: what the first thing do is make sure. I can call func.__module__ without raising an error.
The if is checked left to right, since it is an or, the func.__module__.startswith will be called only if the attribute is there. then I check that the function comes from jax or our modules. 

In general, it is a good idea to get rid of this check